catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,156 @@
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import org.apache.commons.lang3.Validate;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.federation.FederationException;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+/**
+ * Implements iterator. Acts like a decorator.
+ * NOTE: This class is not thread safe. Do not share an object between multiple threads.
+ * Create a new instance by passing a QueryRequestImpl and a CatalogFramework to the constructor.
+ * The paginator will inspect the query's start index and desired page size.
+ * Calling next() returns a list of Result objects in the desired page size.
+ * If the page size is not evenly divisible by the number of results,
+ * then the last page will be smaller than the desired page size.
+ */
+public class QueryResultPaginator implements Iterator<List<Result>> {
+
+    private final CatalogFramework catalogFramework;
+
+    private final QueryRequestImpl queryRequestCopy;
+
+    private int currentIndex;
+
+    private int pageSize;
+
+    //Set to true when a query returns zero results.
+    private boolean exhausted;
+
+    private BlockingQueue<Result> buffer;
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest) {
+        Validate.notNull(catalogFramework);
+        Validate.notNull(queryRequest);
+        this.catalogFramework = catalogFramework;
+        this.queryRequestCopy = clone(queryRequest);
+        pageSize = queryRequestCopy.getQuery()
+                .getPageSize();
+        currentIndex = queryRequest.getQuery()
+                .getStartIndex();
+        Validate.inclusiveBetween(1,
+                Integer.MAX_VALUE,
+                currentIndex,
+                "Query request start index must be greater than zero");
+        exhausted = false;
+        buffer = new LinkedBlockingDeque<>();
+    }
+
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (exhausted && buffer.isEmpty()) {
+            return false;
+        }
+
+        if (!exhausted) {
+            fetchNext();
+        }
+
+        return !buffer.isEmpty();
+    }
+
+    /**
+     * Catalog Framework queried for results until current page is full of results
+     * or endIndex is reached.
+     *
+     * @return Collection of results
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
```

Comment:
❗️ This call to `hasNext()`, unless I'm mistaken, will double-load the queue if the calling code invokes 
```java
hasNext();
next();
```
in succession.